File: plugins/vitest-agent-observability/src/judges/runner.ts
/**
 * Core judge runner: builds prompt, calls LLM once, parses result, records judgments.
 *
 * Sync target:
 *   plugins/pytest-agent-observability/src/pytest_agent_observability/judges/runner.py
 */

import { recordJudgment } from "../collector.js";
import { buildPrompt } from "./prompt.js";
import type { Rubric } from "./rubrics.js";

export interface LLMClient {
  evaluate(prompt: string): Promise<string>;
}

export interface JudgeInput {
  response: string;
  context?: string | string[];
  systemPrompt?: string;
  taskInstructions?: string;
  conversationHistory?: Array<{ role: string; content: string }>;
}

export interface CriterionResult {
  name: string;
  score: number;
  reason: string;
  verdict: "pass" | "fail" | "maybe";
  threshold: number;
}

export async function evaluate({
  rubrics,
  input,
  llm,
  threshold = 0.7,
}: {
  rubrics: Rubric[];
  input: JudgeInput;
  llm: LLMClient;
  threshold?: number;
}): Promise<Record<string, CriterionResult>> {
  const prompt = buildPrompt(rubrics, input);
  const raw = await llm.evaluate(prompt);

  const parsed = parseJson(raw);
  if (parsed === null) {
    recordJudgment({
      intent: "judge_failed",
      name: "judge_failed",
      verdict: "fail",
      reasoning: raw,
    });
    return {};
  }

  const results: Record<string, CriterionResult> = {};
  for (const rubric of rubrics) {
    const entry = parsed[rubric.name];
    if (typeof entry !== "object" || entry === null) continue;
    const score = parseFloat(String(entry.score ?? 0));
    const reason = String(entry.reason ?? "");
    const verdict = score >= threshold ? "pass" : "fail";
    recordJudgment({
      intent: rubric.name,
      name: rubric.name,
      verdict,
      reasoning: reason,
      score,
      threshold,
    });
    results[rubric.name] = { name: rubric.name, score, reason, verdict, threshold };
  }

  return results;
}

function parseJson(raw: string): Record<string, any> | null {
  try {
    const obj = JSON.parse(raw);
    if (typeof obj === "object" && obj !== null && !Array.isArray(obj)) {
      return obj;
    }
  } catch {
    // fall through to regex extraction
  }

  const match = /\{.*\}/s.exec(raw);
  if (match) {
    try {
      const obj = JSON.parse(match[0]);
      if (typeof obj === "object" && obj !== null && !Array.isArray(obj)) {
        return obj;
      }
    } catch {
      // fall through
    }
  }

  return null;
}

File: plugins/pytest-agent-observability/src/pytest_agent_observability/judges/runner.py
"""Core judge runner: builds prompt, calls LLM once, parses result, records judgments."""

from __future__ import annotations

import json
import re
from typing import Any

from ..collector import _record_judgment
from .prompt import build_prompt
from .rubrics import Rubric
from .types import JudgeInput, LLMClient

__all__ = ["JudgeInput", "LLMClient", "evaluate"]


def evaluate(
    rubrics: list[Rubric],
    input: JudgeInput,
    llm: LLMClient,
    threshold: float = 0.7,
) -> dict[str, dict[str, Any]]:
    """Call the LLM once for all rubrics, parse JSON result, record judgments.

    Returns a map of rubric_name -> {"score": float, "reason": str, "verdict": str}.
    On total parse failure returns {} and records a single judge_failed judgment.
    """
    prompt = build_prompt(rubrics, input)
    raw = llm.evaluate(prompt)

    parsed = _parse_json(raw)
    if parsed is None:
        _record_judgment(
            intent="judge_failed",
            verdict="error",
            reasoning=raw,
        )
        return {}

    results: dict[str, dict[str, Any]] = {}
    for rubric in rubrics:
        entry = parsed.get(rubric.name)
        if not isinstance(entry, dict):
            continue
        score = float(entry.get("score", 0.0))
        reason = str(entry.get("reason", ""))
        verdict = "pass" if score >= threshold else "fail"
        _record_judgment(
            intent=rubric.name,
            verdict=verdict,
            reasoning=reason,
            name=rubric.name,
            score=score,
            threshold=threshold,
        )
        results[rubric.name] = {"score": score, "reason": reason, "verdict": verdict}

    return results


def _parse_json(raw: str) -> dict | None:
    """Try strict JSON parse, then extract first {...} block. Returns None on failure."""
    try:
        obj = json.loads(raw)
        if isinstance(obj, dict):
            return obj
    except json.JSONDecodeError:
        pass

    match = re.search(r"\{.*\}", raw, re.DOTALL)
    if match:
        try:
            obj = json.loads(match.group())
            if isinstance(obj, dict):
                return obj
        except json.JSONDecodeError:
            pass

    return None
